src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,332 @@
+
+            foreach (var notary in notaries) await GAS.Mint(engine, notary.EncodePoint(true).ToScriptHash(), singleReward, false);
+        }
+
+        protected override void OnManifestCompose(IsHardforkEnabledDelegate hfChecker, uint blockHeight, ContractManifest manifest)
+        {
+            manifest.SupportedStandards = new[] { "NEP-27" };
+        }
+
+        /// <summary>
+        /// Verify checks whether the transaction is signed by one of the notaries and
+        /// ensures whether deposited amount of GAS is enough to pay the actual sender's fee.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="sig">Signature</param>
+        /// <returns>Whether transaction is valid.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        private bool Verify(ApplicationEngine engine, byte[] sig)
+        {
+            var tx = engine.ScriptContainer as Transaction;
+            if (tx?.GetAttribute<NotaryAssisted>() is null) return false;
+            foreach (var signer in tx.Signers)
+            {
+                if (signer.Account == Hash)
+                {
+                    if (signer.Scopes != WitnessScope.None) return false;
+                    break;
+                }
+            }
+            if (tx.Sender == Hash)
+            {
+                if (tx.Signers.Length != 2) return false;
+                var payer = tx.Signers[1].Account;
+                var balance = GetDepositFor(engine.SnapshotCache, payer);
+                if (balance is null || balance.Amount.CompareTo(tx.NetworkFee + tx.SystemFee) < 0) return false;
+            }
+            ECPoint[] notaries = GetNotaryNodes(engine.SnapshotCache);
+            var hash = tx.GetSignData(engine.GetNetwork());
+            return notaries.Any(n => Crypto.VerifySignature(hash, sig, n));
+        }
+
+        /// <summary>
+        /// OnNEP17Payment is a callback that accepts GAS transfer as Notary deposit.
+        /// It also sets the deposit's lock height after which deposit can be withdrawn.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="from">GAS sender</param>
+        /// <param name="amount">The amount of GAS sent</param>
+        /// <param name="data">Deposit-related data: optional To value (treated as deposit owner if set) and Till height after which deposit can be withdrawn </param>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.All)]
+        private void OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash) throw new InvalidOperationException(string.Format("only GAS can be accepted for deposit, got {0}", engine.CallingScriptHash.ToString()));
+            var to = from;
```

Comment:
Move variable a when is required?

Reply:
Could you clarify please? I'm not sure I understand the question. `to` is an optional parameter in this context, so if it's not specified, then we use the same address as `from`.

Reply:
I mean, define the variable when necessary, there are conditions before that dont required the definition earlier

Reply:
Fixed.